src/frontend/src/expr/mod.rs

```diff
@@ -750,9 +750,31 @@ impl ExprImpl {
             true
         } else if let ExprImpl::FunctionCall(f) = self {
             match f.get_expr_type() {
-                ExprType::Add | ExprType::Subtract => {
+                ty @ (ExprType::Add
+                | ExprType::Subtract
+                | ExprType::AtTimeZone
+                | ExprType::CastWithTimeZone) => {
                     let (_, lhs, rhs) = f.clone().decompose_as_binary();
-                    lhs.is_now_offset() && rhs.is_const()
+                    if matches!(ty, ExprType::AtTimeZone | ExprType::CastWithTimeZone)
+                        && rhs
+                            .as_literal()
+                            .and_then(|literal| literal.get_data().as_ref())
+                            .map_or(true, |time_zone| *time_zone != String::from("UTC").into())
+                    {
+                        false
+                    } else {
+                        lhs.is_now_offset() && rhs.is_const()
+                    }
+                }
+                ExprType::Cast => {
+                    let inputs = f.inputs();
+                    f.return_type() == DataType::Timestamptz
+                        && inputs.len() == 1
+                        && matches!(
+                            inputs[0].return_type(),
+                            DataType::Timestamp | DataType::Timestamptz
+                        )
+                        && inputs[0].is_now_offset()
```

Comment:
Could you provide an example expr / query when this branch is useful? Based on my current understanding:
* timestamptz -> timestamptz cast does not exist
* timestamp -> timestamptz cast without knowing timezone is not consistent with the branch above, where we only allow utc.

Reply:
> Could you provide an example expr / query when this branch is useful? Based on my current understanding:
> 
> * timestamptz -> timestamptz cast does not exist
> * timestamp -> timestamptz cast without knowing timezone is not consistent with the branch above, where we only allow utc.

fixed